cellxgene_schema_cli/cellxgene_schema/validate.py

```diff
@@ -0,0 +1,1588 @@
+                        component, columns_def["add_labels"]
+                    )
+
+            # Do it for index that map to columns
+            if "index" in component_def:
+
+                index_def = component_def["index"]
+                if "add_labels" in index_def:
+                    self._check_single_column_availability(
+                        component, index_def["add_labels"]
+                    )
+
+    def _deep_check(self):
+
+        """
+        Perform a "deep" check of the AnnData object using the schema definition. Adds errors to self.errors if any
+
+        :rtype None
+        """
+
+        # Checks that adata is fully loaded
+        if self.adata.isbacked:
+            raise RuntimeError(
+                "adata is loaded in backed mode, validation requires anndata to be loaded in memory"
+            )
+
+        # Checks that reserved columns are not used
+        self._check_column_availability()
+
+        # Checks sparsity
+        self._validate_sparsity()
+
+        # Checks each component
+        for component, component_def in self.schema_def["components"].items():
+
+            # Skip if component does not exist: only useful for adata.raw.var
+            if _getattr_anndata(self.adata, component) is None:
+                if "required" in component_def:
+                    self.errors.append(
+                        f"'{component}' is missing from adata and it's required"
+                    )
+                continue
+
+            if component_def["type"] == "dataframe":
+                self._validate_dataframe(component)
+            elif component_def["type"] == "dict":
+                dictionary = getattr(self.adata, component)
+                self._validate_dict(dictionary, component, component_def)
+            elif component_def["type"] == "embedding_dict":
+                self._validate_embedding_dict()
+            else:
+                raise ValueError(f"Unexpected component type '{component['type']}'")
+
+        # Checks for raw only if there are no errors, because it depends on the
+        # existence of adata.obs["assay_ontology_term_id"] and adata.obs["X_normalization"]
+        if not self.errors and "raw" in self.schema_def:
+            self._validate_raw()
+
+    def validate_adata(self, h5ad_path: Union[str, bytes, os.PathLike] = None) -> bool:
+
+        """
+        Validates adata
+
+        :params Union[str, bytes, os.PathLike] h5ad_path: path to h5ad to validate, if None it will try to validate
+        from self.adata
+
+        :return True if successful validation, False otherwise
+        :rtype bool
+        """
+
+        # Re-start errors in case a new h5ad is being validated
+        self.errors = []
+
+        if h5ad_path:
+            self._read_h5ad(h5ad_path)
+
+        # Fetches schema def from anndata if schema version is not found in AnnData, this fails
+        self._set_schema_def()
+
+        if not self.errors:
```

Comment:
This if is because `self._set_schema_def()` _can_ add errors _if_ the schema version is not supported. I'm not sure I follow your suggestion.

Reply:
ah, I see. I figured it was trying to avoid performing checks redundantly if the method was called multiple times; I didn't consider the side-effect of `_set_schema_def()`.